Dialog tabindex (#31791)
mdn/content: files/en-us/web/html/element/dialog/index.md

## Usage notes

- HTML {{HTMLElement("form")}} elements can be used to close a dialog box if they have the attribute `method="dialog"` or if the button used to submit the form has [`formmethod="dialog"`](/en-US/docs/Web/HTML/Element/input#formmethod) set. When a `<form>` within a `<dialog>` is submitted via the `dialog` method, the dialog box closes, the states of the form controls are saved but not submitted, and the {{domxref("HTMLDialogElement.returnValue", "returnValue")}} property gets set to the value of the button that was activated.

> - The CSS {{cssxref('::backdrop')}} pseudo-element can be used to style the backdrop of a modal dialog, which is displayed behind the `<dialog>` element when the dialog is displayed using the {{domxref("HTMLDialogElement.showModal()")}} method. For example, this pseudo-element can obfuscate the inert content behind the modal dialog.

Reviewer: ```suggestion
- The CSS {{cssxref('::backdrop')}} pseudo-element can be used to style the backdrop of a modal dialog, which is displayed behind the `<dialog>` element when the dialog is displayed using the {{domxref("HTMLDialogElement.showModal()")}} method. For example, this pseudo-element could be used to blur, darken, or otherwise obfuscate the inert content behind the modal dialog.
```

Reviewer: I just felt that only saying "obfuscate" made the sentence a little harder to understand than providing some examples of what you might do to achieve that.

> - The [`autofocus`](/en-US/docs/Web/HTML/Global_attributes/autofocus) attribute should be added to the element the user is expected to interact with immediately upon opening a modal dialog. If no element involves immediate interaction, while the `autofocus` attribute can be added to the `<dialog>` element itself, it is recommended to provide focus to the close button in the dialog.

Reviewer: ```suggestion
- The [`autofocus`](/en-US/docs/Web/HTML/Global_attributes/autofocus) attribute should be added to the element the user is expected to interact with immediately upon opening a modal dialog. If no other element involves more immediate interaction, it is recommended to add `autofocus` to the close button inside the dialog, or the dialog itself if the user is expected to click/activate it to dismiss.
```

Reviewer: Not 100% sure about the last bit, but I felt this structure was a little more readable, and I liked the idea of incidentally letting the reader know you can autofocus the dialog itself, while providing some kind of use case.

> - Do not add the `tabindex` property to the `<dialog>` element as it is not interactive and does not receive focus. The dialog's contents, including the close button contained in the dialog, can receive focus and be interactive.

Reviewer: ```suggestion
- Do not add the `tabindex` property to the `<dialog>` element as it is not interactive and does not receive focus. The dialog's contents, including the close button contained in the dialog, can receive focus and be interactive. Therefore, you can consider adding `tabindex` to the dialog contents if appropriate.
```

Reviewer: I'm not sure if this is appropriate really, or if you want to encourage readers to use `tabindex` if not absolutely necessary. But, as it was, it felt like you were implying this without saying it, and I feel it is good to be clear and explicit.

Reviewer: Also, now you've added this note, can you just get rid of the big red warning, or maybe link it to this note in some way? I wonder if readers will scan the page, see the big red warning, and still miss the rationale you've added down here.

Author: I prefer not to encourage using tabindex. I had written a sentence about "use semantic HTML elements that are interactive, and tabindex if you can't" but decided this wasn't the spot to discuss semantic elements. I am adding this because of the "why" about the warning. Not to replace it. I'll like to "usage notes". I can't link to the list item directly

Reviewer: Sure, totally understood.

This example demonstrates the [`returnValue`](/en-US/docs/Web/API/HTMLDialogElement/returnValue) of the `<dialog>` element and how to close a modal dialog by using a form. By default, the `returnValue` is the empty string or the value of the button that submits the form within the `<dialog>` element, if there is one.

This example opens a modal dialog when the "Show the dialog" button is activated. The dialog contains a form with a {{HTMLElement("select")}} and two {{HTMLElement("button")}} elements, which default to `type="submit"`. An event listener updates the value of the "Confirm" button when the select option changes. If the "Confirm" button is activated to close the dialog, the current value of the button is the return value. If the dialog is closed by pressing the "Cancel" button, the `returnValue` is `cancel`.

> When the dialog is closed, the return value is displayed under the "Show the dialog" button. If the dialog is closed by pressing the <kbd>Esc</kbd> key, the `return value` is not updated, and the `close` event doesn't occur, so the text in the {{HTMLElement("output")}} is not updated.

Reviewer: ```suggestion
When the dialog is closed, the return value is displayed under the "Show the dialog" button. If the dialog is closed by pressing the <kbd>Esc</kbd> key, the return value is not updated, and the `close` event doesn't occur, so the text in the {{HTMLElement("output")}} is not updated.
```
